internal/vault/credentials/aws.go

```diff
@@ -0,0 +1,220 @@
+	"strconv"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/hashicorp/go-secure-stdlib/awsutil"
+	secretsv1alpha1 "github.com/hashicorp/vault-secrets-operator/api/v1alpha1"
+	"github.com/hashicorp/vault-secrets-operator/internal/consts"
+)
+
+const (
+	AWSAnnotationRole            = "eks.amazonaws.com/role-arn"
+	AWSAnnotationAudience        = "eks.amazonaws.com/audience"
+	AWSAnnotationTokenExpiration = "eks.amazonaws.com/token-expiration"
+	AWSDefaultAudience           = "sts.amazonaws.com"
+	AWSDefaultTokenExpiration    = int64(86400)
+)
+
+var _ CredentialProvider = (*AWSCredentialProvider)(nil)
+
+type AWSCredentialProvider struct {
+	authObj           *secretsv1alpha1.VaultAuth
+	providerNamespace string
+	uid               types.UID
+}
+
+func (l *AWSCredentialProvider) GetNamespace() string {
+	return l.providerNamespace
+}
+
+func (l *AWSCredentialProvider) GetUID() types.UID {
+	return l.uid
+}
+
+func (l *AWSCredentialProvider) Init(ctx context.Context, client ctrlclient.Client, authObj *secretsv1alpha1.VaultAuth, providerNamespace string) error {
+	l.authObj = authObj
+	l.providerNamespace = providerNamespace
+
+	if l.authObj.Spec.AWS.AWSCredsRef != "" {
+		// If AWSCredsRef is not empty, get the secret and read the creds from
+		// there, use the secret UID as l.uid
+		key := ctrlclient.ObjectKey{
+			Namespace: l.providerNamespace,
+			Name:      l.authObj.Spec.AWS.AWSCredsRef,
+		}
+		credsSecret, err := getSecret(ctx, client, key)
+		if err != nil {
+			return err
+		}
+		l.uid = credsSecret.UID
+	} else if l.authObj.Spec.AWS.IRSAServiceAccount != "" {
+		// Otherwise if the IRSA ref is not empty, read the service account, and
+		// use the service account UID as l.uid
+		key := ctrlclient.ObjectKey{
+			Namespace: l.providerNamespace,
+			Name:      l.authObj.Spec.AWS.IRSAServiceAccount,
+		}
+		irsaServiceAccount, err := getSA(ctx, client, key)
+		if err != nil {
+			return err
+		}
+		l.uid = irsaServiceAccount.UID
+	} else {
+		// At this point either the node role or the instance profile will be
+		// used for credentials, so just generate a new UID
+		l.uid = uuid.NewUUID()
+	}
+
+	return nil
+}
+
+func (l *AWSCredentialProvider) GetCreds(ctx context.Context, client ctrlclient.Client) (map[string]interface{}, error) {
+	logger := log.FromContext(ctx)
+	credsSecret := &corev1.Secret{}
+	irsaToken := ""
+	var irsaConfig *IRSAConfig
+
+	if l.authObj.Spec.AWS.AWSCredsRef != "" {
+		var err error
+		key := ctrlclient.ObjectKey{
+			Namespace: l.providerNamespace,
+			Name:      l.authObj.Spec.AWS.AWSCredsRef,
+		}
+		credsSecret, err = getSecret(ctx, client, key)
+		if err != nil {
+			logger.Error(err, "Failed to get secret", "secret_name", l.authObj.Spec.AWS.AWSCredsRef)
+			return nil, err
+		}
+	}
+	if l.authObj.Spec.AWS.IRSAServiceAccount != "" {
```

Comment:
nit: To match the logic in `Init` and avoid any potential errors, which I think would be irrelevant?

```suggestion
	} else if l.authObj.Spec.AWS.IRSAServiceAccount != "" {
```

Reply:
Ok, I think that should work.